crates/wasm-encoder/src/component/types.rs

```diff
@@ -0,0 +1,574 @@
+use super::{ComponentSection, SectionId, TypeRef};
+use crate::{encoders, ValType};
+
+const TYPE_INSTANCE: u8 = 0x7f;
+const TYPE_MODULE: u8 = 0x7e;
+const TYPE_FUNCTION: u8 = 0x7d;
+const TYPE_ADAPTER_FUNCTION: u8 = 0x7c;
+
+const COMPOUND_INTERFACE_TYPE_LIST: u8 = 0x7b;
+const COMPOUND_INTERFACE_TYPE_RECORD: u8 = 0x7a;
+const COMPOUND_INTERFACE_TYPE_VARIANT: u8 = 0x79;
+const COMPOUND_INTERFACE_TYPE_TUPLE: u8 = 0x78;
+const COMPOUND_INTERFACE_TYPE_FLAGS: u8 = 0x77;
+const COMPOUND_INTERFACE_TYPE_ENUM: u8 = 0x76;
+const COMPOUND_INTERFACE_TYPE_UNION: u8 = 0x75;
+const COMPOUND_INTERFACE_TYPE_OPTIONAL: u8 = 0x74;
+const COMPOUND_INTERFACE_TYPE_EXPECTED: u8 = 0x73;
+const COMPOUND_INTERFACE_TYPE_NAMED: u8 = 0x72;
+
+const EXPECTED_OK: u8 = 0x00;
+const EXPECTED_OK_ERR: u8 = 0x01;
+
+const INTERFACE_TYPE_BOOL: u8 = 0x71;
+const INTERFACE_TYPE_S8: u8 = 0x70;
+const INTERFACE_TYPE_U8: u8 = 0x6f;
+const INTERFACE_TYPE_S16: u8 = 0x6e;
+const INTERFACE_TYPE_U16: u8 = 0x6d;
+const INTERFACE_TYPE_S32: u8 = 0x6c;
+const INTERFACE_TYPE_U32: u8 = 0x6b;
+const INTERFACE_TYPE_S64: u8 = 0x6a;
+const INTERFACE_TYPE_U64: u8 = 0x69;
+const INTERFACE_TYPE_F32: u8 = 0x68;
+const INTERFACE_TYPE_F64: u8 = 0x67;
+const INTERFACE_TYPE_CHAR: u8 = 0x66;
+const INTERFACE_TYPE_STRING: u8 = 0x65;
+const INTERFACE_TYPE_UNIT: u8 = 0x64;
+
+const INSTANCE_DEFINITION_KIND: u8 = 0x00;
+const MODULE_DEFINITION_KIND: u8 = 0x01;
+const FUNCTION_DEFINITION_KIND: u8 = 0x02;
+const TABLE_DEFINITION_KIND: u8 = 0x03;
+const MEMORY_DEFINITION_KIND: u8 = 0x04;
+const GLOBAL_DEFINITION_KIND: u8 = 0x05;
+
+const ALIAS_TYPE_INSTANCE_EXPORT: u8 = 0x00;
+const ALIAS_TYPE_OUTER: u8 = 0x01;
+
+const OUTER_ALIAS_MODULE: u8 = 0x01;
+const OUTER_ALIAS_TYPE: u8 = 0x06;
+
+const INSTANCE_TYPEDEF_TYPE: u8 = 0x01;
+const INSTANCE_TYPEDEF_ALIAS: u8 = 0x05;
+const INSTANCE_TYPEDEF_EXPORT: u8 = 0x06;
+
+const MODULE_TYPEDEF_TYPE: u8 = 0x01;
+const MODULE_TYPEDEF_ALIAS: u8 = 0x05;
+const MODULE_TYPEDEF_EXPORT: u8 = 0x06;
+const MODULE_TYPEDEF_IMPORT: u8 = 0x02;
+
+/// Represents a definition kind.
+///
+/// This is used in aliases to specify the kind of the definition.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DefinitionKind {
+    /// The definition is an instance.
+    Instance,
+    /// The definition is a module.
+    Module,
+    /// The definition is a function.
+    Function,
+    /// The definition is a table.
+    Table,
+    /// The definition is a memory.
+    Memory,
+    /// The definition is a global.
+    Global,
+}
+
+impl DefinitionKind {
+    pub(crate) fn encode(&self, bytes: &mut Vec<u8>) {
+        match self {
+            Self::Instance => bytes.push(INSTANCE_DEFINITION_KIND),
```

Comment:
Instead of a separate constant for these could the byte value be the discriminant for the `enum`?

Reply:
I'll make that change.